File: apps/rokid-glass-probe/app/src/main/java/com/realitymemory/glassprobe/WearStateReceiver.java
package com.realitymemory.glassprobe;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

public class WearStateReceiver extends BroadcastReceiver {
    static final String ACTION_TAKE_STATUS_CHANGED = "com.rokid.sprite.ACTION_TAKE_STATUS_CHANGED";
    static final String ACTION_LEG_STATUS_CHANGED = "com.rokid.sprite.ACTION_LEG_STATUS_CHANGED";
    private static final String EXTRA_TAKE_STATE = "glasses_take_state";

    @Override
    public void onReceive(Context context, Intent intent) {
        String detail = describeIntent(intent);
        ProbeLog.append(context, "WEAR_RECEIVER", detail);

        if (ACTION_TAKE_STATUS_CHANGED.equals(intent.getAction())) {
            String state = intent.getStringExtra(EXTRA_TAKE_STATE);
            if ("1".equals(state)) {
                sendService(context, CaptureForegroundService.ACTION_WEAR_DETECTED);
            } else if ("0".equals(state)) {
                sendService(context, CaptureForegroundService.ACTION_STOP);
            }
        }
    }

    static String describeIntent(Intent intent) {
        StringBuilder builder = new StringBuilder();
        builder.append("action=").append(intent.getAction());
        Bundle extras = intent.getExtras();
        if (extras != null) {
            for (String key : extras.keySet()) {
                Object value = extras.get(key);
                builder.append(", ").append(key).append("=").append(value);
            }
        }
        return builder.toString();
    }

    private void sendService(Context context, String action) {
        Intent serviceIntent = new Intent(context, CaptureForegroundService.class);
        serviceIntent.setAction(action);
        try {
            context.startForegroundService(serviceIntent);
        } catch (Exception e) {
            ProbeLog.append(context, "WEAR_SERVICE_START_FAILED", e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }
}

File: apps/reality-memory-glasses/app/src/main/java/com/realitymemory/glasses/runtime/WearStateReceiver.kt
package com.realitymemory.glasses.runtime

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import com.realitymemory.glasses.MainActivity

class WearStateReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            Intent.ACTION_BOOT_COMPLETED -> startMonitor(context)

            ACTION_TAKE_STATUS_CHANGED -> {
                val worn = intent.getStringExtra(EXTRA_TAKE_STATE) == "1"
                startRuntime(context, worn, "TAKE_STATUS")
                if (worn) showDisclosure(context)
            }

            ACTION_LEG_STATUS_CHANGED -> {
                val unfolded = intent.getStringExtra(EXTRA_LEG_STATE) == "1"
                if (!unfolded) startRuntime(context, false, "LEGS_FOLDED")
            }
        }
    }

    private fun startMonitor(context: Context) {
        ContextCompat.startForegroundService(
            context,
            Intent(context, RealityRuntimeService::class.java)
                .setAction(RealityRuntimeService.ACTION_ARM_MONITOR),
        )
    }

    private fun startRuntime(context: Context, worn: Boolean, source: String) {
        val serviceIntent = Intent(context, RealityRuntimeService::class.java)
            .setAction(RealityRuntimeService.ACTION_WEAR_CHANGED)
            .putExtra(RealityRuntimeService.EXTRA_WORN, worn)
            .putExtra(RealityRuntimeService.EXTRA_WEAR_SOURCE, source)
        ContextCompat.startForegroundService(context, serviceIntent)
    }

    private fun showDisclosure(context: Context) {
        runCatching {
            context.startActivity(
                Intent(context, MainActivity::class.java)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_SINGLE_TOP)
                    .putExtra(MainActivity.EXTRA_OPENED_FROM_WEAR, true),
            )
        }
    }

    companion object {
        const val ACTION_TAKE_STATUS_CHANGED = "com.rokid.sprite.ACTION_TAKE_STATUS_CHANGED"
        const val ACTION_LEG_STATUS_CHANGED = "com.rokid.sprite.ACTION_LEG_STATUS_CHANGED"
        private const val EXTRA_TAKE_STATE = "glasses_take_state"
        private const val EXTRA_LEG_STATE = "glasses_leg_state"
    }
}
